R/merge_sce_list.R

```diff
@@ -0,0 +1,199 @@
+#' Merge a list of SCEs as preparation for formal integration
+#'
+#' This function takes a named (ideally by a form of library ID) list of
+#'  SingleCellExperiment (SCE) objects and merges them into one SCE object.
+#'  All names in the colData slot must match across provided SCE objects.
+#'  The resulting colData slot will contain information in the given batch column
+#'  to differentiate originating SCE objects. Only features (genes) that are present
+#'  in all provided SCE objects will be retained. To preserve original feature-level
+#'  information, all final rowData slot column names will be appended with the
+#'  given SCE's name, as `<column_name>-<sce_name>`. Currently any present altExps
+#'  are not retained.
+#'
+#' @param sce_list A named list of SingleCellExperiment objects
+#' @param batch_column A character value giving the resulting colData column name
+#'  to differentiate originating SingleCellExperiment objects. Often these values
+#'  are library IDs. Default value is `library_id`.
+#' @param barcode_column A character value giving the colData column name that
+#'  stores cell barcodes. If this column does not yet exist it will be created. Default
+#'  is `barcode`.
+#'  The name of the column to be created (or not if it
+#'   already exists) in the colData slot to store cell barcodes
+#' @param retain_coldata_cols A vector of colData columns which should be retained
+#'  in the main experiment of the final merged SCE object.
+#' @param preserve_rowdata_cols A vector of column names that appear in originating
+#'  SingleCellExperiment objects' rowData slots which should not be renamed with
+#'  the given SingleCellExperiment object name. These are generally columns which are
+#'  not specific to the given library's preparation or statistics. For example,
+#'  such an array might contain items like "Gene", "ensembl_ids", "gene_symbol", etc.
+#'
+#' @return A SingleCellExperiment object containing all SingleCellExperiment objects
+#'   present in the inputted list
+#' @export
+#'
+#' @import SingleCellExperiment
+merge_sce_list <- function(sce_list = list(),
+                           batch_column = "library_id",
+                           barcode_column = "barcode",
+                           retain_coldata_cols = c("sum",
```

Comment:
Retaining barcode explicitly rather than implicitly.
```suggestion
                           retain_coldata_cols = c("barcode,
                                                   "sum",
```

Reply:
I accepted this but I think I'd rather change it a bit. Since the `barcode_column` itself is an argument, I don't want to hardcode a default value here. Instead I'm going to just update `retain_coldata_cols` in the code to add in whatever the `barcode_column` value is.

Reply:
I'd rather not do that. My thought is that the code should not modify the `retain_coldata_cols` at all. If the user is changing the default `barcode_column`, then they may need to change `retain_coldata_cols`, and the function could check for that (it already kind of does, in that if the `"barcode"` column does not exist, it will complain), but adding a column to retain that the user did not ask for seems like an odd behavior.

Do need to add a missing quote to my suggestion though, at the least!

Reply:
@jashapiro Ok that said I'm getting confused with some of the feedback here on barcodes, so I'm going to try and suss it out here:

1.  Per your comment we should only "retain the `barcode` column if explicitly asked", I should definitely undo this accepted change. This means I should not, as I wrote, "update `retain_coldata_cols` in the code to add in whatever the `barcode_column` value is." If the user wants it they have to tell us `retain_coldata_cols`.
2. We need _all_ SCEs to either have or not have a barcode column, so if some are missing barcodes but some are not, we can't not add anything at all. We have to either make `NA` barcode column(s) where missing or remove barcode column(s) where present. We certainly do not want to be _removing_ any cell-identifying information though!
3. So, be it a barcode specifically or some other unique identifier (without presuming what exactly this information is), we need some kind of cell IDing information no matter what. An overall argument `cell_id_column` might be the way to go (rather than `barcode_column` which has an expectation about the type of information), and in the docs explain these _could_ be barcodes. We could then handle this column in a couple different ways:
- Simply require it. Users have to tell us unique cell ID'ing information. This avoids the trouble of what if some SCEs have cell IDs and others don't. I think this scenario is less likely than SCEs being consistent, but it certainly doesn't seem _un_likely that this may happen.
- If missing in any SCE, fill it in with `colnames(sce)`. This _may_ mean different forms of IDs if some SCEs have barcodes and some don't if some were originally missing.

I think we're on the same page now? Any preference for requiring cell info vs. creating it? My preference is to require it.

Reply:
Agree mostly with point 1, except to say that the change I suggest was to the default value for the argument. So in my mind accepting the default value is "explicitly asked", but I can see that might not be a general view!

Agree on point 2, and yes, we would want to add `NA` values. Which is what we do for other retained columns without full data, correct?

The `cell_id_column` that you suggest is, I think, the `colnames` and I don't think we need to add a separate column for it; However, this is the one where we may need to be careful. It should either be the current  `colnames` + `batch_id` or `barcode_id` + `batch_id`. I don't think it should really matter either way and if they are not 100% consistent, I don't think that is a problem, so my initial thought was to use the barcode if it exists and the colname if it doesn't. (Later thought: we might reverse the order in that ID so that batch comes first)

Reply:
> Agree mostly with point 1, except to say that the change I suggested was to the default value for the argument. So in my mind accepting the default value is "explicitly asked", but I can see that might not be a general view!

Good point! I'm going to flesh this out alongside the barcode situation, keep reading...

> Agree on point 2, and yes, we would want to add `NA` values. Which is what we do for other retained columns without full data, correct?

Correct! But now I'm circling back to disagreeing with myself and maybe advocating for NOT filling in with `NA`.🙃 

I think we have to require barcode consistency among their input SCEs. They either all have a barcode, or none have a barcode. If it varies among SCEs, we should error out. Any variation will prevent consistency among `cell_id_column` values. For example, if only some SCEs have barcodes, then we'll end up in either of these not great circumstances:
  -  cell IDs are `{batch}-NA` a `{batch}-{barcode} ` naming scheme (noting I like the idea of batch first for this value). This is not great.
  - Some SCEs have a `cell_id_column` scheme of `{batch}-{barcode}` but others may have scheme of `{batch}-{rowname}`, and that inconsistency is also not great.


What I am leaning to is:
- If there is a barcode column, create a `cell_id_column` (which would not be in the provided retained columns vector since we are making it) as `{batch}-{barcode}`. Keep the barcode column as a default as you originally suggested (although don't hardcode the `barcode_column` into the args).
- If there is NO barcode column, create a `cell_id_column` as `{batch}-{rowname}`.
  - If there are no rownames, `stop()` 
  - Do not, however, create a barcode column since these will be `NA`s all around and that doesn't seem like it helps anybody.

Reply:
Note some of this TBD code I pushed for different ways we can handle the barcode column..
https://github.com/AlexsLemonade/scpcaTools/blob/261dfe8458024a0d71e900f5fa3af24103fff459/R/merge_sce_list.R#L71-L85

https://github.com/AlexsLemonade/scpcaTools/blob/261dfe8458024a0d71e900f5fa3af24103fff459/R/merge_sce_list.R#L163-L170

I haven't added anything for `cell_id_column` yet.

Reply:
> * Some SCEs have a `cell_id_column` scheme of `{batch}-{barcode}` but others may have scheme of `{batch}-{rowname}`, and that inconsistency is also not great.

I don't think this is necessarily a problem! The cell ids are pretty arbitrary, so one random string vs. another is not going to show up as a problem. But if we are worried about it, I would just can all the worry and use `{batch}-{rowname}`.

We would then remove everything around `barcode_column`, including the option. Treat `barcode` just like any other retained column, filling in `NA` values when missing (but make it one of the default retained columns, for our own ease of use and good example).

Reply:
> * If there is a barcode column, create a `cell_id_column` (which would not be in the provided retained columns vector since we are making it) as `{batch}-{barcode}`. Keep the barcode column as a default as you originally suggested (although don't hardcode the `barcode_column` into the args).

If we decide to remove the `barcode_column` logic, or even if we don't, it might make sense to make a `cell_id` column and include it _always_. (Do I contradict myself, then I contradict myself!) My reason is that this will allow us to trace back to prior (unmerged) SCE objects should the need arise, without parsing the merged cell_id/colnames.

> * If there is NO barcode column, create a `cell_id_column` as `{batch}-{rowname}`.
>   
>   * If there are no rownames, `stop()`
 This should never occur, as rownames of some kind are always required by DataFrame.

Reply:
Ok I'm going for it - 
- barcode is going to be a normal column, and I'll remove the argument. 
- cell_id will be made and will always be retained; whether they ask or not they get it. 100000% correct that we need something to trace back, and I prefer it being a cell_id as `{batch}-{rowname}`.




> This should never occur, as rownames of some kind are always required by DataFrame. If just numbers, we would still want to include those.

Good to know!